base_layer/core/src/base_node/service/blockchain_state/service.rs

```diff
@@ -0,0 +1,211 @@
+};
+use futures::{
+    channel::mpsc,
+    future,
+    future::Either,
+    stream::{Fuse, FuturesUnordered},
+    StreamExt,
+    TryFutureExt,
+};
+use log::*;
+use std::{future::Future, ops::Bound};
+use tari_comms::protocol::rpc::__macro_reexports::BoxFuture;
+use tokio::future::poll_fn;
+
+const LOG_TARGET: &str = "c::base_node::blockchain_state_service";
+
+pub struct BlockchainStateService<B> {
+    db: BlockchainDatabase<B>,
+    requests: Fuse<mpsc::Receiver<BlockchainStateRequest>>,
+}
+
+impl<B: BlockchainBackend + 'static> BlockchainStateService<B> {
+    pub fn new(db: BlockchainDatabase<B>, requests: mpsc::Receiver<BlockchainStateRequest>) -> Self {
+        Self {
+            db,
+            requests: requests.fuse(),
+        }
+    }
+
+    #[inline]
+    fn db(&self) -> BlockchainDatabase<B> {
+        self.db.clone()
+    }
+
+    pub async fn run(mut self) {
+        // This "work queue" holds all the pending tasks. The reason this is a good fit for this service is that this
+        // service is only for calls to the database wrapped in some ergonomics. All async_db calls spawn a
+        // blocking task. So the work queue doesn't actually have to do much work itself, it is simply waiting
+        // for the blocking task to notify it when it is complete, so polling many of these tasks sequentially is very
+        // lightweight.
+        let mut work_queue = FuturesUnordered::new();
+
+        loop {
+            futures::select! {
+                // Add each request to a "work queue"
+                request = self.requests.select_next_some() => {
+                    let work = self.handle_request(request);
+                    work_queue.push(work);
+                },
+                // Process the work queue
+                _ = work_queue.select_next_some() => { },
+            }
+        }
+    }
+
+    fn handle_request(&mut self, request: BlockchainStateRequest) -> BoxFuture<'static, ()> {
+        use BlockchainStateRequest::*;
+        trace!(target: LOG_TARGET, "Got request: {:?}", request);
+        match request {
+            GetBlocks((start, end), reply) => {
+                Box::pin(reply_or_cancel(reply, Self::fetch_blocks(self.db(), start, end)))
+            },
+            GetHeaders((start, end), reply) => {
+                Box::pin(reply_or_cancel(reply, Self::fetch_headers(self.db(), start, end)))
+            },
+            GetHeaderByHeight(height, reply) => Box::pin(reply_or_cancel(
+                reply,
+                async_db::fetch_header(self.db(), height).map_err(Into::into),
+            )),
+            GetChainMetadata(reply) => Box::pin(reply_or_cancel(
+                reply,
+                async_db::get_chain_metadata(self.db()).map_err(Into::into),
+            )),
+            FindFirstMatchingHeader(hashes, reply) => Box::pin(reply_or_cancel(
+                reply,
+                Self::find_first_matching_header(self.db(), hashes),
```

Comment:
In my opinion this should start sending blocks or headers back immediately. The rational is that a malicious node could advertise a false tip height, then on request, it can return a valid header and then ghost you on the next call.

If the sender must immediately provide a chain, it must have created a fake chain in order to respond to this request. This reduces the attack surface significantly.

Reply:
How about this?

```proto
message FindChainSplitRequest {
  repeated bytes block_hashes = 1;
  uint64 count = 2;
}

message FindChainSplitResponse {
  // An ordered list of headers starting from next header after the matching hash, up until `FindChainSplitRequest::count`
  repeated tari.core.BlockHeader headers = 1;
  // The index of the hash that matched from `FindChainSplitRequest::block_hashes`. This value could also be used to know how far back a split occurs.
  uint32 found_hash_index = 2;
}

```